improve multi-core performance of qmax_pool2d

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* #68328
* #72575
* #71903
* #69667
* #69601
* #69600
* #69599
* __->__ #69598
* #69517

Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

mingfeima committed:
improve multi-core performance of qmax_pool2d

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353159).

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168 We have a few int8-related PRs targeting 1.11. And this is one of them. Could you prioritize to review these PRs? Thanks.

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

jerryzh168 commented:
sure, I'm asking @dzdang from our team to review the stack. Also please provide benchmark numbers for single core and multi-core perf for all the PRs to make sure we don't regress the performance

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353159).

mingfeima commented:
### Updates of this PR
improves `qmax_pool` multi-core performance on channels last memory format. similar to #69517.

* **contiguous memory format**: the original kernel has nested parallel loops (which is typically not recommended from performance perspective). The outer loop parallels on {N} and the inner loop parallels on {C}. This will be problematic if N ! = 1 but N is a small number, e.g. N = 2 will run on 2 cores even if more cores are available. Since the omp wrapper in pytorch `at::parallel_for` will force the inner loop go sequential. But this PR doesn't address this issue, this part stay untouched.
* **channels last memory format**: originally parallels on {N} and go sequentially on {H, W}, vectorized on {C}. The problem here is that in case of single batch inference (when N=1), it will go sequentially even on a multi-core CPU. This PR parallels on {N, H, W} and vectorized on {C}. e.g. given an input shape of {1, 64, 112, 112}, the original kernel run sequentially and the new kernel will run parallelly.


### Benchmarking

* **short tag**: `python -m pt.qpool_test`
* **long tag**: `python  -m pt.qpool_test --tag_filter long`

CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, dual sockets, 20 cores per sockets.
Unit: us

#### 1.a) single core run on short tag
this PR has no effect on cases from short tag since they are all `contigTrue` (which means nchw memory format).

name | before | after | speedup
-- | -- | -- | --
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.qint32 | 6.748 | 6.889 | 97.95%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.qint8 | 6.69 | 6.868 | 97.41%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.quint8 | 6.572 | 6.658 | 98.71%


#### 1.b) single socket run on short tag
this PR has no effect on cases from short tag since they are all `contigTrue` (which means nchw memory format).

name | before | after | speedup
-- | -- | -- | --
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.qint32 | 8.958 | 8.501 | 105.38%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.qint8 | 8.962 | 8.606 | 104.14%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N2_contigTrue_dtypetorch.quint8 | 7.733 | 8.33 | 92.83%



#### 2.a) single core run on long tag
this PR has no effect on single core run

name | before | after | speedup
-- | -- | -- | --
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 6.456 | 6.672 | 96.76%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 6.442 | 6.544 | 98.44%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 6.525 | 6.837 | 95.44%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 6.503 | 6.48 | 100.35%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N1_contigFalse_dtypetorch.quint8 | 28.575 | 30.049 | 95.09%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N1_contigTrue_dtypetorch.quint8 | 33.986 | 31.692 | 107.24%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N4_contigFalse_dtypetorch.quint8 | 94.231 | 98.429 | 95.73%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N4_contigTrue_dtypetorch.quint8 | 98.692 | 96.983 | 101.76%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 166.285 | 165.882 | 100.24%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 3938.395 | 3761.552 | 104.70%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 619.953 | 628.994 | 98.56%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 14106.49 | 14095.64 | 100.08%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 33.066 | 34.274 | 96.48%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 999.343 | 1000.612 | 99.87%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 165.289 | 166.787 | 99.10%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 3781.727 | 3782.676 | 99.97%


#### 2.b) single socket run on long tag
this PR improves `contigFalse` (which means nhwc memory format) on multi-core, for `contigTrue` (nchw), there is no effect.

name | before | after | speedup
-- | -- | -- | --
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 7.956 | 8.242 | 96.53%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 7.762 | 7.691 | 100.92%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 7.989 | 8.257 | 96.75%
QMaxPool2dBenchmark_C1_H3_W3_k(3,3)_s(1,1)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 7.8 | 8.171 | 95.46%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N1_contigFalse_dtypetorch.quint8 | 34.334 | 13.116 | 261.77%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N1_contigTrue_dtypetorch.quint8 | 21.862 | 19.431 | 112.51%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N4_contigFalse_dtypetorch.quint8 | 105.872 | 20.422 | 518.42%
QMaxPool2dBenchmark_C3_H64_W64_k(3,3)_s(2,2)_p(1,1)_N4_contigTrue_dtypetorch.quint8 | 39.152 | 41.465 | 94.42%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 125.375 | 24.305 | 515.84%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 275.642 | 286.398 | 96.24%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 447.247 | 62.024 | 721.09%
QMaxPool2dBenchmark_C64_H224_W224_k(2,2)_s(2,2)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 4329.375 | 4304.987 | 100.57%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N1_contigFalse_dtypetorch.quint8 | 34.593 | 14.218 | 243.30%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N1_contigTrue_dtypetorch.quint8 | 68.488 | 70.878 | 96.63%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N4_contigFalse_dtypetorch.quint8 | 107.482 | 17.18 | 625.62%
QMaxPool2dBenchmark_C256_H56_W56_k(2,2)_s(2,2)_p(0,0)_N4_contigTrue_dtypetorch.quint8 | 1160.208 | 1165.604 | 99.54%

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353159).

mingfeima commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -1359,87 +1359,85 @@ void qmaxpool_2d_nhwc_kernel(
+      data_index_init(begin, b, nBatch, row, oH, col, oW);
+
+      for (const auto i : c10::irange(begin, end)) {
+        auto* i_p = reinterpret_cast<scalar_t::underlying*>(idata + b * iW * iH * iC);
+        auto* o_p = reinterpret_cast<scalar_t::underlying*>(odata + i * iC);
```
@dzdang, as I understand this kernel doesn't allocate space when iterating each input HW plane when updating corresponding value on the output index.

the above are calculating input and start output index of each (n, h, w) and start input index of (n).

usually on `channels last` memory format we can have two types of implementation for pooling kernels, since we are vectoring on `channels`:
1. iterate each input lane of channels at a time and store the immediate values on output lane of channels, advantage is memory access is contiguous, disadvantage is need to extra writes to output. when scalar_t != acc_scalar_t, need to allocate extra buffer to keep the immediate values on acc_scalar_t. for example the current [fp32](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L99) and [bf16](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L247) kernels use this approach, the temp buffer is thread local and reused across output (n, h, w).
2. iterate each input lane of channels at a block at a time, this kernel uses this approach, the block is 4 * vec_length. advantage is that no extra write to output since it is using __m256i to store the immediate values, disadvantage is the memory access pattern is non-contiguous (when moving to next block, the stride is C).

mingfeima replied:
by the way, this PR is only change the kernel from sequential to parallel, noting else is changed.

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

mingfeima committed:
Update on "improve multi-core performance of qmax_pool2d"


Differential Revision: [D33353159](https://our.internmc.facebook.com/intern/diff/D33353159)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353159).
